File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/API/Manager/Inventory/Right/MissionInventory.java
package jp.aoichaan0513.A_TosoGame_Live.API.Manager.Inventory.Right;

import jp.aoichaan0513.A_TosoGame_Live.API.Interfaces.IMission;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.MissionManager;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.*;

public class MissionInventory {

    public static final String missionTitle = ChatColor.DARK_GRAY + "> " + ChatColor.BOLD + "ホーム" + ChatColor.RESET + ChatColor.DARK_GRAY + " > " + ChatColor.BOLD + "ミッションアプリ";
    public static final String tutatuHintTitle = ChatColor.DARK_GRAY + "> " + ChatColor.BOLD + "ホーム" + ChatColor.RESET + ChatColor.DARK_GRAY + " > " + ChatColor.BOLD + "ミッションアプリ (通達・ヒント)";
    public static final String endTitle = ChatColor.DARK_GRAY + "> " + ChatColor.BOLD + "ホーム" + ChatColor.RESET + ChatColor.DARK_GRAY + " > " + ChatColor.BOLD + "ミッションアプリ (終了)";

    private static HashMap<Integer, IMission> missionList = new HashMap<>();

    private static int missionAmount = 1;
    private static int tutatuHintAmount = 1;
    private static int endAmount = 1;

    private static int missionSlot = 18;
    private static int tutatuHintSlot = 18;
    private static int endSlot = 18;

    private static Inventory missionInventory = Bukkit.createInventory(null, 9 * 6, missionTitle);
    private static Inventory tutatuHintInventory = Bukkit.createInventory(null, 9 * 6, tutatuHintTitle);
    private static Inventory endInventory = Bukkit.createInventory(null, 9 * 6, endTitle);

    private static final ItemFlag[] itemFlags = new ItemFlag[]{ItemFlag.HIDE_ATTRIBUTES, ItemFlag.HIDE_ENCHANTS};

    public static void reset() {
        missionAmount = 1;
        tutatuHintAmount = 1;
        endAmount = 1;

        missionSlot = 18;
        tutatuHintSlot = 18;
        endSlot = 18;

        resetInventory();
        resetMissions();
    }

    public static void resetInventory() {
        missionInventory = Bukkit.createInventory(null, 9 * 6, missionTitle);
        tutatuHintInventory = Bukkit.createInventory(null, 9 * 6, tutatuHintTitle);
        endInventory = Bukkit.createInventory(null, 9 * 6, endTitle);
        return;
    }

    public static Inventory getInventory(MissionManager.MissionBookType type) {
        if (type == MissionManager.MissionBookType.MISSION) {
            Inventory inv = missionInventory;

            ItemStack missionStack = new ItemStack(Material.RED_CONCRETE);
            ItemMeta missionMeta = missionStack.getItemMeta();
            missionMeta.addItemFlags(itemFlags);
            missionMeta.addEnchant(Enchantment.DURABILITY, 1, true);
            missionMeta.setDisplayName("" + ChatColor.RED + ChatColor.BOLD + ChatColor.UNDERLINE + "ミッション");
            missionMeta.setLore(Collections.singletonList(ChatColor.GRAY + "クリックして" + ChatColor.RED + ChatColor.BOLD + ChatColor.UNDERLINE + "現在実行中のミッションリスト" + ChatColor.GRAY + "を開きます。"));
            missionStack.setItemMeta(missionMeta);


            for (int i = 0; i < 3; i++)
                inv.setItem(i, missionStack);

            ItemStack tutatuHintStack = new ItemStack(Material.YELLOW_CONCRETE);
            ItemMeta tutatuHintMeta = tutatuHintStack.getItemMeta();
            tutatuHintMeta.addItemFlags(itemFlags);
            tutatuHintMeta.setDisplayName("" + ChatColor.YELLOW + ChatColor.BOLD + ChatColor.UNDERLINE + "通達・ヒント");
            tutatuHintMeta.setLore(Collections.singletonList(ChatColor.GRAY + "クリックして" + ChatColor.YELLOW + ChatColor.BOLD + ChatColor.UNDERLINE + "通達・ヒントリスト" + ChatColor.GRAY + "を開きます。"));
            tutatuHintStack.setItemMeta(tutatuHintMeta);

            for (int i = 3; i < 6; i++)
                inv.setItem(i, tutatuHintStack);

            ItemStack endStack = new ItemStack(Material.LIME_CONCRETE);
            ItemMeta endMeta = endStack.getItemMeta();
            endMeta.addItemFlags(itemFlags);
            endMeta.setDisplayName("" + ChatColor.GREEN + ChatColor.BOLD + ChatColor.UNDERLINE + "終了したミッション");
            endMeta.setLore(Collections.singletonList(ChatColor.GRAY + "クリックして" + ChatColor.GREEN + ChatColor.BOLD + ChatColor.UNDERLINE + "終了したミッションのリスト" + ChatColor.GRAY + "を開きます。"));
            endStack.setItemMeta(endMeta);

            for (int i = 6; i < 9; i++)
                inv.setItem(i, endStack);

            ItemStack grayStack = new ItemStack(Material.BLACK_STAINED_GLASS_PANE);
            ItemMeta grayMeta = grayStack.getItemMeta();
            grayMeta.addItemFlags(itemFlags);
            grayMeta.setDisplayName(ChatColor.BOLD + "");
            grayStack.setItemMeta(grayMeta);

            for (int i = 9; i < 18; i++)
                inv.setItem(i, grayStack);
            for (int i = 45; i < 49; i++)
                inv.setItem(i, grayStack);
            for (int i = 50; i < 54; i++)
                inv.setItem(i, grayStack);

            ItemStack homeStack = new ItemStack(Material.WHITE_STAINED_GLASS_PANE);
            ItemMeta homeMeta = homeStack.getItemMeta();
            homeMeta.addItemFlags(itemFlags);
            homeMeta.setDisplayName("" + ChatColor.GOLD + ChatColor.BOLD + ChatColor.UNDERLINE + "ホーム");
            homeMeta.setLore(Collections.singletonList(ChatColor.YELLOW + "クリックして" + ChatColor.GOLD + ChatColor.BOLD + ChatColor.UNDERLINE + "ホーム画面" + ChatColor.RESET + ChatColor.YELLOW + "を表示します。"));
            homeStack.setItemMeta(homeMeta);

            inv.setItem(49, homeStack);

            return inv;
        } else if (type == MissionManager.MissionBookType.TUTATU || type == MissionManager.MissionBookType.HINT) {
            Inventory inv = tutatuHintInventory;

            ItemStack missionStack = new ItemStack(Material.RED_CONCRETE);
            ItemMeta missionMeta = missionStack.getItemMeta();
            missionMeta.addItemFlags(itemFlags);
            missionMeta.setDisplayName("" + ChatColor.RED + ChatColor.BOLD + ChatColor.UNDERLINE + "ミッション");
            missionMeta.setLore(Collections.singletonList(ChatColor.GRAY + "クリックして" + ChatColor.RED + ChatColor.BOLD + ChatColor.UNDERLINE + "現在実行中のミッションリスト" + ChatColor.GRAY + "を開きます。"));
            missionStack.setItemMeta(missionMeta);

            for (int i = 0; i < 3; i++)
                inv.setItem(i, missionStack);

            ItemStack tutatuHintStack = new ItemStack(Material.YELLOW_CONCRETE);
            ItemMeta tutatuHintMeta = tutatuHintStack.getItemMeta();
            tutatuHintMeta.addItemFlags(itemFlags);
            tutatuHintMeta.addEnchant(Enchantment.DURABILITY, 1, true);
            tutatuHintMeta.setDisplayName("" + ChatColor.YELLOW + ChatColor.BOLD + ChatColor.UNDERLINE + "通達・ヒント");
            tutatuHintMeta.setLore(Collections.singletonList(ChatColor.GRAY + "クリックして" + ChatColor.YELLOW + ChatColor.BOLD + ChatColor.UNDERLINE + "通達・ヒントリスト" + ChatColor.GRAY + "を開きます。"));
            tutatuHintStack.setItemMeta(tutatuHintMeta);

            for (int i = 3; i < 6; i++)
                inv.setItem(i, tutatuHintStack);

            ItemStack endStack = new ItemStack(Material.LIME_CONCRETE);
            ItemMeta endMeta = endStack.getItemMeta();
            endMeta.addItemFlags(itemFlags);
            endMeta.setDisplayName("" + ChatColor.GREEN + ChatColor.BOLD + ChatColor.UNDERLINE + "終了したミッション");
            endMeta.setLore(Collections.singletonList(ChatColor.GRAY + "クリックして" + ChatColor.GREEN + ChatColor.BOLD + ChatColor.UNDERLINE + "終了したミッションのリスト" + ChatColor.GRAY + "を開きます。"));
            endStack.setItemMeta(endMeta);

            for (int i = 6; i < 9; i++)
                inv.setItem(i, endStack);

            ItemStack grayStack = new ItemStack(Material.BLACK_STAINED_GLASS_PANE);
            ItemMeta grayMeta = grayStack.getItemMeta();
            grayMeta.addItemFlags(itemFlags);
            grayMeta.setDisplayName(ChatColor.BOLD + "");
            grayStack.setItemMeta(grayMeta);

            for (int i = 9; i < 18; i++)
                inv.setItem(i, grayStack);
            for (int i = 45; i < 49; i++)
                inv.setItem(i, grayStack);
            for (int i = 50; i < 54; i++)
                inv.setItem(i, grayStack);

            ItemStack homeStack = new ItemStack(Material.WHITE_STAINED_GLASS_PANE);
            ItemMeta homeMeta = homeStack.getItemMeta();
            homeMeta.addItemFlags(itemFlags);
            homeMeta.setDisplayName("" + ChatColor.GOLD + ChatColor.BOLD + ChatColor.UNDERLINE + "ホーム");
            homeMeta.setLore(Collections.singletonList(ChatColor.YELLOW + "クリックして" + ChatColor.GOLD + ChatColor.BOLD + ChatColor.UNDERLINE + "ホーム画面" + ChatColor.RESET + ChatColor.YELLOW + "を表示します。"));
            homeStack.setItemMeta(homeMeta);

            inv.setItem(49, homeStack);

            return inv;
        } else {
            Inventory inv = endInventory;

            ItemStack missionStack = new ItemStack(Material.RED_CONCRETE);
            ItemMeta missionMeta = missionStack.getItemMeta();
            missionMeta.addItemFlags(itemFlags);
            missionMeta.setDisplayName("" + ChatColor.RED + ChatColor.BOLD + ChatColor.UNDERLINE + "ミッション");
            missionMeta.setLore(Collections.singletonList(ChatColor.GRAY + "クリックして" + ChatColor.RED + ChatColor.BOLD + ChatColor.UNDERLINE + "現在実行中のミッションリスト" + ChatColor.GRAY + "を開きます。"));
            missionStack.setItemMeta(missionMeta);

            for (int i = 0; i < 3; i++)
                inv.setItem(i, missionStack);

            ItemStack tutatuHintStack = new ItemStack(Material.YELLOW_CONCRETE);
            ItemMeta tutatuHintMeta = tutatuHintStack.getItemMeta();
            tutatuHintMeta.addItemFlags(itemFlags);
            tutatuHintMeta.setDisplayName("" + ChatColor.YELLOW + ChatColor.BOLD + ChatColor.UNDERLINE + "通達・ヒント");
            tutatuHintMeta.setLore(Collections.singletonList(ChatColor.GRAY + "クリックして" + ChatColor.YELLOW + ChatColor.BOLD + ChatColor.UNDERLINE + "通達・ヒントリスト" + ChatColor.GRAY + "を開きます。"));
            tutatuHintStack.setItemMeta(tutatuHintMeta);

            for (int i = 3; i < 6; i++)
                inv.setItem(i, tutatuHintStack);

            ItemStack endStack = new ItemStack(Material.LIME_CONCRETE);
            ItemMeta endMeta = endStack.getItemMeta();
            endMeta.addItemFlags(itemFlags);
            endMeta.addEnchant(Enchantment.DURABILITY, 1, true);
            endMeta.setDisplayName("" + ChatColor.GREEN + ChatColor.BOLD + ChatColor.UNDERLINE + "終了したミッション");
            endMeta.setLore(Collections.singletonList(ChatColor.GRAY + "クリックして" + ChatColor.GREEN + ChatColor.BOLD + ChatColor.UNDERLINE + "終了したミッションのリスト" + ChatColor.GRAY + "を開きます。"));
            endStack.setItemMeta(endMeta);

            for (int i = 6; i < 9; i++)
                inv.setItem(i, endStack);

            ItemStack grayStack = new ItemStack(Material.BLACK_STAINED_GLASS_PANE);
            ItemMeta grayMeta = grayStack.getItemMeta();
            grayMeta.addItemFlags(itemFlags);
            grayMeta.setDisplayName(ChatColor.BOLD + "");
            grayStack.setItemMeta(grayMeta);

            for (int i = 9; i < 18; i++)
                inv.setItem(i, grayStack);
            for (int i = 45; i < 49; i++)
                inv.setItem(i, grayStack);
            for (int i = 50; i < 54; i++)
                inv.setItem(i, grayStack);

            ItemStack homeStack = new ItemStack(Material.WHITE_STAINED_GLASS_PANE);
            ItemMeta homeMeta = homeStack.getItemMeta();
            homeMeta.addItemFlags(itemFlags);
            homeMeta.setDisplayName("" + ChatColor.GOLD + ChatColor.BOLD + ChatColor.UNDERLINE + "ホーム");
            homeMeta.setLore(Collections.singletonList(ChatColor.YELLOW + "クリックして" + ChatColor.GOLD + ChatColor.BOLD + ChatColor.UNDERLINE + "ホーム画面" + ChatColor.RESET + ChatColor.YELLOW + "を表示します。"));
            homeStack.setItemMeta(homeMeta);

            inv.setItem(49, homeStack);

            return inv;
        }
    }

    public static void setInventory(Inventory inv, MissionManager.MissionBookType type) {
        if (type == MissionManager.MissionBookType.MISSION) {
            missionInventory = inv;
            return;
        } else if (type == MissionManager.MissionBookType.TUTATU || type == MissionManager.MissionBookType.HINT) {
            tutatuHintInventory = inv;
            return;
        } else {
            endInventory = inv;
            return;
        }
    }

    public static void resetMissions() {
        missionList = new HashMap<>();
        return;
    }

    public static void setMissions(HashMap<Integer, IMission> hashMap) {
        missionList = hashMap;
        return;
    }

    public static HashMap<Integer, IMission> getMissions(MissionManager.MissionBookType type) {
        HashMap<Integer, IMission> hashMap = new HashMap<>();

        for (Map.Entry<Integer, IMission> entry : missionList.entrySet()) {
            IMission mission = entry.getValue();

            if (type == MissionManager.MissionBookType.MISSION) {
                if (mission.getType() == MissionManager.MissionBookType.MISSION)
                    hashMap.put(entry.getKey(), mission);
            } else if (type == MissionManager.MissionBookType.TUTATU || type == MissionManager.MissionBookType.HINT) {
                if (mission.getType() == MissionManager.MissionBookType.TUTATU || mission.getType() == MissionManager.MissionBookType.HINT)
                    hashMap.put(entry.getKey(), mission);
            } else {
                if (mission.getType() == MissionManager.MissionBookType.END_MISSION
                        || mission.getType() == MissionManager.MissionBookType.END_TUTATU || mission.getType() == MissionManager.MissionBookType.END_HINT)
                    hashMap.put(entry.getKey(), mission);
            }
        }
        return hashMap;
    }

    private static void addItem(String name, String[] lore, MissionManager.MissionBookType type, Material material) {
        int amount = getAmount(type);
        int slot = getSlot(type);

        Inventory inv = getInventory(type);

        ItemStack itemStack = new ItemStack(material, amount);
        ItemMeta itemMeta = itemStack.getItemMeta();
        itemMeta.setDisplayName(name);
        itemMeta.setLore(Arrays.asList(lore));
        itemStack.setItemMeta(itemMeta);

        inv.setItem(slot, itemStack);

        setInventory(inv, type);

        setSlot(slot + 1, type);
        setAmount(amount + 1, type);
        return;
    }

    public static void removeItem(MissionManager.MissionBookType type, int i) {
        Inventory inv = getInventory(type);
        inv.remove(inv.getItem(i));
        return;
    }

    public static ItemStack getItem(MissionManager.MissionBookType type, int i) {
        return getInventory(type).getItem(i);
    }

    public static ItemStack getItem(MissionManager.MissionBookType type, Material material, int amount) {
        for (ItemStack itemStack : getInventory(type).getContents())
            if (itemStack.getType() == material && itemStack.getAmount() == amount)
                return itemStack;
        return null;
    }

    public static void addMission(String title, List<String> descriptions, MissionManager.MissionBookType type, Material material) {
        int amount = getAmount(type);
        int slot = getSlot(type);

        if (slot < 45) {
            addItem(ChatColor.BOLD + title, new String[]{ChatColor.YELLOW + "続きはこちら"}, type, material);
            missionList.put((type.getId() + (amount * slot)), new IMission(title, descriptions, type, material, amount));
            return;
        }
        return;
    }

    public static void endMission() {
        Inventory missionInv = getInventory(MissionManager.MissionBookType.MISSION);
        Inventory endInv = getInventory(MissionManager.MissionBookType.END_MISSION);

        for (int i = 18; i < 44; i++) {
            int amount = getAmount(MissionManager.MissionBookType.END_MISSION);
            int slot = getSlot(MissionManager.MissionBookType.END_MISSION);

            ItemStack itemStack = missionInv.getItem(i);

            if (itemStack == null) continue;

            endInv.addItem(itemStack);
            missionInv.remove(itemStack);

            setSlot(slot + 1, MissionManager.MissionBookType.END_MISSION);
            setAmount(amount + 1, MissionManager.MissionBookType.END_MISSION);
        }

        setInventory(missionInv, MissionManager.MissionBookType.MISSION);
        setInventory(endInv, MissionManager.MissionBookType.END_MISSION);

        setSlot(18, MissionManager.MissionBookType.MISSION);
        return;
    }

    public static void openBook(ItemStack book, Player p) {
        p.openBook(book);
        /*
        int slot = p.getInventory().getHeldItemSlot();
        ItemStack old = p.getInventory().getItem(slot);
        p.getInventory().setItem(slot, book);

        ByteBuf buf = Unpooled.buffer(256);
        buf.setByte(0, (byte) 0);
        buf.writerIndex(1);

        PacketPlayOutCustomPayload packet = new PacketPlayOutCustomPayload(new MinecraftKey("minecraft:book_open"), new PacketDataSerializer(buf));
        ((CraftPlayer) p).getHandle().playerConnection.sendPacket(packet);
        p.getInventory().setItem(slot, old);
        */
    }

    public static int getAmount(MissionManager.MissionBookType type) {
        if (type == MissionManager.MissionBookType.MISSION) {
            return missionAmount;
        } else if (type == MissionManager.MissionBookType.TUTATU || type == MissionManager.MissionBookType.HINT) {
            return tutatuHintAmount;
        } else {
            return endAmount;
        }
    }

    private static void setAmount(int amount, MissionManager.MissionBookType type) {
        if (type == MissionManager.MissionBookType.MISSION) {
            missionAmount = amount;
        } else if (type == MissionManager.MissionBookType.TUTATU || type == MissionManager.MissionBookType.HINT) {
            tutatuHintAmount = amount;
        } else {
            endAmount = amount;
        }
    }

    public static int getSlot(MissionManager.MissionBookType type) {
        if (type == MissionManager.MissionBookType.MISSION) {
            return missionSlot;
        } else if (type == MissionManager.MissionBookType.TUTATU || type == MissionManager.MissionBookType.HINT) {
            return tutatuHintSlot;
        } else {
            return endSlot;
        }
    }

    private static void setSlot(int slot, MissionManager.MissionBookType type) {
        if (type == MissionManager.MissionBookType.MISSION) {
            missionSlot = slot;
        } else if (type == MissionManager.MissionBookType.TUTATU || type == MissionManager.MissionBookType.HINT) {
            tutatuHintSlot = slot;
        } else {
            endSlot = slot;
        }
    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/Inventory/MissionInventory.kt
package jp.aoichaan0513.A_TosoGame_Live.Inventory

import jp.aoichaan0513.A_TosoGame_Live.API.Interfaces.IMission
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.PlayerManager
import jp.aoichaan0513.A_TosoGame_Live.Mission.MissionManager
import jp.aoichaan0513.A_TosoGame_Live.Utils.ItemUtil
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.Player
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.BookMeta

class MissionInventory {
    companion object {

        val missionTitle = "${ChatColor.DARK_GRAY}> ${ChatColor.BOLD}ホーム${ChatColor.RESET}${ChatColor.DARK_GRAY} > ${ChatColor.BOLD}ミッションアプリ"
        val tutatuHintTitle = "${ChatColor.DARK_GRAY}> ${ChatColor.BOLD}ホーム${ChatColor.RESET}${ChatColor.DARK_GRAY} > ${ChatColor.BOLD}ミッションアプリ (通達・ヒント)"
        val endTitle = "${ChatColor.DARK_GRAY}> ${ChatColor.BOLD}ホーム${ChatColor.RESET}${ChatColor.DARK_GRAY} > ${ChatColor.BOLD}ミッションアプリ (終了)"


        fun getInventory(type: MissionManager.MissionType): Inventory {
            val inv = when (type) {
                MissionManager.MissionType.TUTATU_HINT -> Bukkit.createInventory(null, 9 * 6, tutatuHintTitle)
                MissionManager.MissionType.END -> Bukkit.createInventory(null, 9 * 6, endTitle)
                else -> Bukkit.createInventory(null, 9 * 6, missionTitle)
            }

            val itemStackBorder = ItemStack(Material.BLACK_STAINED_GLASS_PANE)
            val itemMetaBorder = itemStackBorder.itemMeta!!
            itemMetaBorder.addItemFlags(*ItemUtil.itemFlags)
            itemMetaBorder.setDisplayName("${ChatColor.BOLD}")
            itemStackBorder.itemMeta = itemMetaBorder
            for (i in 9..17)
                inv.setItem(i, itemStackBorder)
            for (i in 45..53)
                inv.setItem(i, itemStackBorder)


            val itemStackHome = ItemStack(Material.WHITE_STAINED_GLASS_PANE)
            val itemMetaHome = itemStackHome.itemMeta!!
            itemMetaHome.addItemFlags(*ItemUtil.itemFlags)
            itemMetaHome.setDisplayName("${ChatColor.GOLD}${ChatColor.BOLD}${ChatColor.UNDERLINE}ホーム")
            itemMetaHome.lore = listOf("${ChatColor.YELLOW}クリックして${ChatColor.GOLD}${ChatColor.BOLD}${ChatColor.UNDERLINE}ホーム画面${ChatColor.RESET}${ChatColor.YELLOW}を表示します。")
            itemStackHome.itemMeta = itemMetaHome
            inv.setItem(49, itemStackHome)

            val itemStackMission = ItemStack(Material.RED_CONCRETE)
            val itemMetaMission = itemStackMission.itemMeta!!
            itemMetaMission.addItemFlags(*ItemUtil.itemFlags)
            if (type == MissionManager.MissionType.MISSION)
                itemMetaMission.addEnchant(Enchantment.DURABILITY, 1, true)
            itemMetaMission.setDisplayName("${ChatColor.RED}${ChatColor.BOLD}${ChatColor.UNDERLINE}ミッション")
            itemMetaMission.lore = listOf("${ChatColor.GRAY}クリックして${ChatColor.RED}${ChatColor.BOLD}${ChatColor.UNDERLINE}現在実行中のミッションリスト${ChatColor.RESET}${ChatColor.GRAY}を開きます。")
            itemStackMission.itemMeta = itemMetaMission
            for (i in 0..2)
                inv.setItem(i, itemStackMission)

            val itemStackTutatuHint = ItemStack(Material.YELLOW_CONCRETE)
            val itemMetaTutatuHint = itemStackTutatuHint.itemMeta!!
            itemMetaTutatuHint.addItemFlags(*ItemUtil.itemFlags)
            if (type == MissionManager.MissionType.TUTATU_HINT)
                itemMetaTutatuHint.addEnchant(Enchantment.DURABILITY, 1, true)
            itemMetaTutatuHint.setDisplayName("${ChatColor.YELLOW}${ChatColor.BOLD}${ChatColor.UNDERLINE}通達・ヒント")
            itemMetaTutatuHint.lore = listOf("${ChatColor.GRAY}クリックして${ChatColor.YELLOW}${ChatColor.BOLD}${ChatColor.UNDERLINE}通達・ヒントリスト${ChatColor.RESET}${ChatColor.GRAY}を開きます。")
            itemStackTutatuHint.itemMeta = itemMetaTutatuHint
            for (i in 3..5)
                inv.setItem(i, itemStackTutatuHint)

            val itemStackEnd = ItemStack(Material.LIME_CONCRETE)
            val itemMetaEnd = itemStackEnd.itemMeta!!
            itemMetaEnd.addItemFlags(*ItemUtil.itemFlags)
            if (type == MissionManager.MissionType.END)
                itemMetaEnd.addEnchant(Enchantment.DURABILITY, 1, true)
            itemMetaEnd.setDisplayName("${ChatColor.GREEN}${ChatColor.BOLD}${ChatColor.UNDERLINE}終了したミッション")
            itemMetaEnd.lore = listOf("${ChatColor.GRAY}クリックして${ChatColor.GREEN}${ChatColor.BOLD}${ChatColor.UNDERLINE}終了したミッションのリスト${ChatColor.RESET}${ChatColor.GRAY}を開きます。")
            itemStackEnd.itemMeta = itemMetaEnd
            for (i in 6..8)
                inv.setItem(i, itemStackEnd)

            for (mission in MissionManager.getMissions(type)) {
                val descriptionFirst = mission.descriptions[0]

                val itemStackMissionDetail = ItemStack(mission.material, mission.count)
                val itemMetaMissionDetail = itemStackMissionDetail.itemMeta!!
                itemMetaMissionDetail.addItemFlags(*ItemUtil.itemFlags)
                itemMetaMissionDetail.setCustomModelData(mission.count)

                itemMetaMissionDetail.setDisplayName(when (type) {
                    MissionManager.MissionType.TUTATU_HINT -> "${ChatColor.YELLOW}${ChatColor.BOLD}${ChatColor.UNDERLINE}${mission.title}"
                    MissionManager.MissionType.END -> "${ChatColor.GREEN}${ChatColor.BOLD}${ChatColor.UNDERLINE}${mission.title}"
                    else -> "${ChatColor.RED}${ChatColor.BOLD}${ChatColor.UNDERLINE}${mission.title}"
                })

                itemMetaMissionDetail.lore = if (descriptionFirst.length > 15) {
                    listOf("${ChatColor.RESET}${descriptionFirst.substring(0, 15).replace("\n", " ")}…", "${ChatColor.GRAY}${ChatColor.UNDERLINE}続きはここをクリック")
                } else {
                    listOf("${ChatColor.RESET}${descriptionFirst.replace("\n", " ")}")
                }

                itemStackMissionDetail.itemMeta = itemMetaMissionDetail
                inv.addItem(itemStackMissionDetail)
            }
            return inv
        }

        fun openBook(p: Player, id: Int, missionType: MissionManager.MissionType = MissionManager.MissionType.MISSION) {
            openBook(p, MissionManager.getMission(id, missionType) ?: return)
        }

        fun openBook(p: Player, mission: IMission) {
            val bookStack = ItemStack(Material.WRITTEN_BOOK, 1)
            val bookMeta = bookStack.itemMeta as BookMeta
            bookMeta.title = "逃走中"
            bookMeta.author = "A_TosoGame_Live"
            for (v in mission.descriptions.indices)
                bookMeta.addPage("${if (v == 0) "${ChatColor.DARK_RED}${mission.title}${ChatColor.RESET}\n\n" else ""}${PlayerManager.loadConfig(p).bookForegroundColor.color}${mission.descriptions[v]}")
            bookStack.itemMeta = bookMeta
            openBook(p, bookStack)
        }

        private fun openBook(p: Player, book: ItemStack?) {
            p.openBook(book ?: return)
        }
    }
}